# Додаткове завдання з pandas
Вітаю! В цьому завданні будемо працювати з даними з опитування щодо зарплат ІТ спеціалістів на рексурсі DOU від грудня 2022 року. Впевнена, це буде цікаве дослідження, а заразом і практика pandas :)
Цей аналіз може бути корисним людині, яка обирає напрям роботи в ІТ і хоче зрозуміти тенденції. За якісного виконання цього аналізу, цей проєкт можна сміливо додавати у Ваше портфоліо!

1. Зчитайте дані з файлу `2022_dec_raw.csv`. Увага! Тут необхідно використати аргумент `pd.read_csv` `sep=';'`, бо розділювач в даних - крапка з комою.

In [1]:
import pandas as pd

In [2]:
data_path = '2022_dec_raw.csv'
df = pd.read_csv(data_path, sep=';')

2. Які колонки мають дані?

In [3]:
df.columns

Index(['Timestamp', 'Ви працюєте зараз в ІТ?', 'Де ви зараз живете? ',
       'В якій області ви зараз живете?',
       'Зарплата у $$$ за місяць, лише ставка після сплати податків',
       'Ваша стать', 'Ваш вік', 'Яка у вас освіта?',
       'Чи є у вас науковий ступінь? ', 'Знання англійської мови',
       'Оберіть вашу посаду', 'Ваш тайтл', 'Ваша спеціалізація',
       'Вкажіть вашу спеціалізацію', 'Оберіть вашу спеціалізацію',
       'Ваша посада', 'Оберіть вашу посаду2', 'Ваша посада3', 'Ваша посада4',
       'Ваша посада5', 'Ваша посада6', 'Ваша посада7',
       'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
       'Основна мова програмування', 'Вкажіть вашу основну спеціалізацію',
       'В якій сфері працюєте?', 'Тип компанії',
       'Загальний стаж роботи за спеціальністю'],
      dtype='str')

3. Скільки рядків та колонок мають дані? Скільки пропущених значень має кожна з колонок?

In [4]:
df.shape

(13315, 28)

In [5]:
df.isna().sum()

Timestamp                                                                         0
Ви працюєте зараз в ІТ?                                                           0
Де ви зараз живете?                                                               0
В якій області ви зараз живете?                                                1777
Зарплата у $$$ за місяць, лише ставка після сплати податків                      27
Ваша стать                                                                        0
Ваш вік                                                                           0
Яка у вас освіта?                                                                 0
Чи є у вас науковий ступінь?                                                    111
Знання англійської мови                                                           0
Оберіть вашу посаду                                                               0
Ваш тайтл                                                                   

4. Які колонки мають всі значення заповненими (не мають дожного None значення)? Виведіть програмно список цих колонок.

In [6]:
full_columns = list(df.columns[df.isna().sum() == 0])
full_columns

['Timestamp',
 'Ви працюєте зараз в ІТ?',
 'Де ви зараз живете? ',
 'Ваша стать',
 'Ваш вік',
 'Яка у вас освіта?',
 'Знання англійської мови',
 'Оберіть вашу посаду',
 'Ваш тайтл',
 'В якій сфері працюєте?',
 'Тип компанії',
 'Загальний стаж роботи за спеціальністю']

5. Який тип мають колонки в наборі даних? Які з колонок типу object логічно на ваш погляд перетворити на числові колонки?

In [7]:
df.dtypes

Timestamp                                                                     str
Ви працюєте зараз в ІТ?                                                       str
Де ви зараз живете?                                                           str
В якій області ви зараз живете?                                               str
Зарплата у $$$ за місяць, лише ставка після сплати податків                   str
Ваша стать                                                                    str
Ваш вік                                                                       str
Яка у вас освіта?                                                             str
Чи є у вас науковий ступінь?                                                  str
Знання англійської мови                                                       str
Оберіть вашу посаду                                                           str
Ваш тайтл                                                                     str
Ваша спеціалізац

**Колонки які доцільно перетворити на числові:**

Зарплата у $$$ за місяць, лише ставка після сплати податків 

Ваш вік

Загальний стаж роботи за спеціальністю  


6. Трансформуйте колонки 'Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків' у тип даних float. УВАГА! Перед трансформацією подивіться, які дані містять ці колонки. Аби успішно трансформувати тип даних, треба спочатку привести рядки до того виду, який буде підходити, аби потім рядок зробити типом float.

In [9]:
df['Ваш вік'].unique()

<StringArray>
[  '29',   '36',   '34',   '23',   '33',   '31',   '30',   '24',   '27',
   '17',   '25',   '20',   '26',   '22',   '19',   '21',   '18',   '32',
   '35',   '28',   '37',   '43',   '44',   '39',   '50',   '40',   '38',
   '41',   '45',   '42',   '49',   '46',   '47',   '48',   '55',   '63',
   '58',   '53',   '54',   '51',   '56',   '60',   '16',   '57', '29,5',
   '52',   '14']
Length: 47, dtype: str

In [10]:
df['Зарплата у $$$ за місяць, лише ставка після сплати податків'].unique()

<StringArray>
['4000',    nan, '2500',  '500', '2800', '5500', '1400', '1200',  '650',
 '6000',
 ...
 '1156', '3790',  '958',  '876',  '407',  '271', '3975',  '163', '2165',
 '1930']
Length: 897, dtype: str

In [11]:
df['Ваш вік'] = (
    df['Ваш вік']
    .astype(str)
    .str.strip()
    .str.replace(',', '.', regex=False)
)
df['Ваш вік'] = df['Ваш вік'].astype(float)

df['Ваш вік'].dtypes

dtype('float64')

In [12]:
salary_col = 'Зарплата у $$$ за місяць, лише ставка після сплати податків'

df[salary_col] = (
    df[salary_col]
    .astype(str)
    .str.strip()
    .str.replace(' ', '', regex=False)   # прибрати пробіли
    .str.replace(',', '.', regex=False) 
    .replace(['nan', 'NaN', 'None', ''], pd.NA)
)


df[salary_col] = df[salary_col].astype(float)
df[salary_col].dtypes

dtype('float64')

7. Це завдання з зірочкою! Нам треба транфсформувати колонку Timestamp у тип даних `datetime64[ns]`. Це не так просто як здається на першйий погляд :) Але це реальна ситуація, яка може трапитись при зборі даних в роботі аналітиком.
Я запропоную один зі способів виконати завдання, але ви можете написати по-своєму. Головне - отримати колонку - потрібного типу.
Пропоную наступний алгоритм дій:
    1. Подивитись, які дані містяться в цій колонці. Виведіть перші і останні 5 записів. Чи однаковий формат цих даних? (спойлер: ні, бо інакше все було б просто :))
    2. Нам треба враховуючи різний формат даних перетворити рядок у колонці `Timestamp` на тип `datetime64[ns]`. Як це зробити? Для перетворення рядка на `datetime64` пропоную скористатись `datetime.datetime.strptime`.
    Для цього треба імпортувати бібліотеку `datetime`. Я пропоную написати функцію для обробки кожного окремого рядка, яка враховує довжину рядка і на основі цього застосовує різний формат даних дати.
    Як прописувати формат даних можна переглянути [тут](https://www.w3schools.com/python/python_datetime.asp).
    3. Коли функція для обробки кожного окремого рядка готова, застосуйте її для всіх даних з допомогою методу `apply`.

In [13]:
df['Timestamp'].head()

0    11.30.2022 10:26:37
1    11.30.2022 10:26:39
2    11.30.2022 10:26:58
3    11.30.2022 10:27:31
4    11.30.2022 10:27:34
Name: Timestamp, dtype: str

In [14]:
df['Timestamp'].tail()

13310    01.09.2023 11:01
13311    01.09.2023 11:14
13312    01.09.2023 11:19
13313    01.09.2023 11:28
13314    01.09.2023 11:38
Name: Timestamp, dtype: str

In [15]:
from datetime import datetime

In [16]:
def parse_timestamp(x):
    if pd.isna(x):
        return pd.NaT

    x = str(x).strip()

    try:
        if len(x) == 19:
            return datetime.strptime(x, "%m.%d.%Y %H:%M:%S")
        else:
            return datetime.strptime(x, "%d.%m.%Y %H:%M")
    except:
        return pd.NaT


df['Timestamp'] = df['Timestamp'].apply(parse_timestamp)
df['Timestamp'] = df['Timestamp'].astype('datetime64[ns]')
df['Timestamp'].dtypes

dtype('<M8[ns]')

8. Для колонок типу даних object (які такими лишились після всіх наших трансформацій) виведіть кількість унікальних значень.

In [17]:
df.select_dtypes(include='str').nunique()

Ви працюєте зараз в ІТ?                                                          5
Де ви зараз живете?                                                              2
В якій області ви зараз живете?                                                 23
Ваша стать                                                                       2
Яка у вас освіта?                                                                5
Чи є у вас науковий ступінь?                                                     3
Знання англійської мови                                                          5
Оберіть вашу посаду                                                             22
Ваш тайтл                                                                       10
Ваша спеціалізація                                                               8
Вкажіть вашу спеціалізацію                                                       4
Оберіть вашу спеціалізацію                                                      21
Ваша

9. Для колонок типу даних object виведіть топ 10 найбільш частих значень з кількістю рядків, в яких це значення зустрічається. Зробіть це для кожної колонки окремо. Наприклад, в циклі.  
Що цікавого бачите? Які висновки можна зробити? Що особисто Вас вразило?

In [19]:
object_cols = df.select_dtypes(include='str').columns

for col in object_cols:
    print(f'\n==========')
    print(df[col].value_counts().head(10))


Ви працюєте зараз в ІТ?
Так, працюю full-time      12643
Ні, тимчасово не працюю      380
Так, працюю part-time        152
Так, я фрилансер             115
Інше                          25
Name: count, dtype: int64

Де ви зараз живете? 
В Україні                                                      11587
За кордоном (переїхали через війну, але планую повернутися)     1728
Name: count, dtype: int64

В якій області ви зараз живете?
Київ чи область                4411
Львів чи область               2353
Дніпро чи область               751
Івано-Франківськ чи область     478
Вінниця чи область              447
Одеса чи область                443
Харків чи область               355
Полтава чи область              252
Черкаси чи область              232
Хмельницький чи область         230
Name: count, dtype: int64

Ваша стать
Чоловік    10221
Жінка       3094
Name: count, dtype: int64

Яка у вас освіта?
Вища (бакалавр, спеціаліст, магістр)    10645
Дві вищі                                 1

Попри війну, переважна більшість айтівців залишаються в Україні. Співвідношення чоловіків і жінок приблизно 3.3 до 1. ІТ індустрія все ще залишається переважно чоловічою. Cпеціалістів Data Science та ML у списку лише 288 проти майже 7000 розробників. Це показує, що сфера Data — це все ще досить вузька ніша. Тільки 12% респондентів мають рівень Advanced.

10. Виведіть медіанну (медіана - це значення, менше якого є рівно 50% вибірки і більше якого є теж рівно 50% вибірки)
ЗП (Зарплата у $$$) для кожного рівня знання англійської. Які висновки можна зробити?

In [62]:
df.groupby('Знання англійської мови')[
    'Зарплата у $$$ за місяць, лише ставка після сплати податків'
].median().sort_values(ascending=False)

Знання англійської мови
Advanced              3500.0
Upper-Intermediate    3300.0
Intermediate          2125.0
Pre-Intermediate      1404.0
Elementary            1155.0
Name: Зарплата у $$$ за місяць, лише ставка після сплати податків, dtype: float64

Щоб стати високооплачуваним фахівцем, недостатньо лише технічних навичок — потрібно вільно володіти мовою. Найбільший стрибок у зарплаті відбувається при переході від Intermediate до Upper-Intermediate.

11. Є така штука як сортування значень категоріальної змінної. Нам це зараз стане в нагоді, адже рівні англійської можна посортувати.
Давайте відсотруємо рівні англ. і тайтли. Просто запустіть код нижче і проаналізуйте, що він робить.

In [20]:
df['Знання англійської мови'] = pd.Categorical(df['Знання англійської мови'],
                      categories=['Elementary','Pre-Intermediate','Intermediate','Upper-Intermediate','Advanced'],
                      ordered=True)

df['Ваш тайтл'] = pd.Categorical(df['Ваш тайтл'],
                      categories=['Intern/Trainee', 'Junior', 'Немає тайтлу', 'Middle', 'Senior', 'Tech Lead', 'Team Lead', 'Architect', 'Head', 'Manager'],
                      ordered=True)

Цей код створює категоріальні змінні для рівня англійської мови та тайтлів. Це дозволяє задати логічну ієрархію категорій та коректно виконувати сортування і агрегацію даних.

12. Зробіть зводну таблицю, де стовпці - це 'Ваш тайтл', рядки - 'Знання англійської мови', а значення - медіанна ЗП. Які висновки можна зробити?
Тут ще для зручності можна зробити градієнтне забарвлення і відформатувати значення для легшого перегляду.

In [25]:
pivot = pd.pivot_table(
    df,
    values='Зарплата у $$$ за місяць, лише ставка після сплати податків',
    index='Знання англійської мови',
    columns='Ваш тайтл',
    aggfunc='median'
)

styled_pivot = pivot.style \
    .background_gradient(cmap='YlGnBu', axis=None) \
    .format("{:.0f}", na_rep="-") \
    .highlight_null(color='lightgrey') \
    .set_caption("Медіанна ЗП в розрізі тайтлів та рівня англійської мови")

styled_pivot

Ваш тайтл,Intern/Trainee,Junior,Немає тайтлу,Middle,Senior,Tech Lead,Team Lead,Architect,Head,Manager
Знання англійської мови,,,,,,,,,,
Elementary,500,600,550,1500,2500,5000,2100,-,1500,1120
Pre-Intermediate,400,800,934,1825,3500,4000,2850,2900,2400,1225
Intermediate,400,900,1300,2200,4100,5000,3800,5450,3500,2100
Upper-Intermediate,485,1000,1600,2500,4750,5650,4655,6400,4000,3600
Advanced,500,1000,2075,2500,4750,6000,5000,7000,5000,4000


Рівень англійської є одни із ключових факторів, який впливає на рівень зарплати. Чим вищий рівень англійської, тим вищу зарплату можна отримати.
Tech Lead з Elementary ($5000) виглядає як аномалія або дуже специфічний випадок.

13. Виведіть результат методу describe() по колонцы розміру ЗП для топ 5 за частотою зустрічання в даних областей (В якій області ви зараз живете?). Зробіть висновки, де вигідніше жити з точки зору ЗП і чи суттєва різниця на Ваш погляд для прийняття рішення про переїзд?

In [66]:
top_regions = df['В якій області ви зараз живете?'].value_counts().head(5).index
df_top = df[df['В якій області ви зараз живете?'].isin(top_regions)]
salary_col = 'Зарплата у $$$ за місяць, лише ставка після сплати податків'

df_top.groupby('В якій області ви зараз живете?')[salary_col].describe()

,count,mean,std,min,25%,50%,75%,max
В якій області ви зараз живете?,,,,,,,,
Івано-Франківськ чи область,478.0,2705.257322,1905.673385,200.0,1200.0,2372.5,3775.0,12436.0
Вінниця чи область,447.0,2888.306085,1902.668331,100.0,1300.0,2612.0,4100.0,11675.0
Дніпро чи область,751.0,2582.584554,1864.711284,100.0,1025.0,2000.0,3600.0,12000.0
Київ чи область,4411.0,3141.810209,2039.859712,80.0,1500.0,2800.0,4500.0,16000.0
Львів чи область,2353.0,3121.441751,2102.721295,150.0,1375.0,2700.0,4500.0,14000.0


Київ та Львів є лідерами і демонструють найвищі медіанні зарплати. Різниця між ними мінімальна, що робить ці два регіони  рівноцінними за рівнем доходів в ІТ. Вінниця посідає третє місце з медіаною $2612, випереджаючи Івано-Франківськ та Дніпро.

Фінансова вигода від переїзду зараз не є вирішальною, адже зараз є можливість працювати віддалено.

Переїзд може бути фінансово виправданим з Дніпра до Києва чи Львова, де різниця в медіані складає 700–800 доларів на місяць. Це вагома сума, яка покриває витрати на релокацію та краще житло.

14. Як зростає зарплата з роками? Для цього нам треба відформатувати колонку "Загальний стаж роботи за спеціальністю" аналогічно до того, як ми це робили для колонки 'Знання англійської мови' і потім вивести медіанні значення для кількості років стажу.

In [67]:
df['Загальний стаж роботи за спеціальністю'].unique()

<StringArray>
[          '8 років',            '3 роки',            '4 роки',
          'Пів року',           '6 років',          '1,5 року',
             '1 рік',            '2 роки',           '5 років',
          '10 років', 'Менше як 3 місяці',          '11 років',
           '7 років', '15 і більше років',          '3 місяці',
          '12 років',           '9 років',          '14 років',
          '13 років']
Length: 19, dtype: str

In [68]:
df['Загальний стаж роботи за спеціальністю'] = pd.Categorical(
    df['Загальний стаж роботи за спеціальністю'],
    categories=[
        'Менше як 3 місяці',
        '3 місяці',
        'Пів року',
        '1 рік',
        '1,5 року',
        '2 роки',
        '3 роки',
        '4 роки',
        '5 років',
        '6 років',
        '7 років',
        '8 років',
        '9 років',
        '10 років',
        '11 років',
        '12 років',
        '13 років',
        '14 років',
        '15 і більше років'
    ],
    ordered=True
)

In [71]:
salary_col = 'Зарплата у $$$ за місяць, лише ставка після сплати податків'

df.groupby('Загальний стаж роботи за спеціальністю')[salary_col].median()


Загальний стаж роботи за спеціальністю
Менше як 3 місяці     800.0
3 місяці              525.0
Пів року              700.0
1 рік                 950.0
1,5 року             1200.0
2 роки               1700.0
3 роки               2450.0
4 роки               3000.0
5 років              3500.0
6 років              4000.0
7 років              4250.0
8 років              4700.0
9 років              4500.0
10 років             4600.0
11 років             5000.0
12 років             4750.0
13 років             4500.0
14 років             5050.0
15 і більше років    4500.0
Name: Зарплата у $$$ за місяць, лише ставка після сплати податків, dtype: float64

Найшвидше зростання відбувається у період 1–5 роки: У цей період зарплата зростає майже в 4 рази. Після 8 років досвіду медіана вперше сягає свого максимуму ($4700). Далі ми спостерігаємо певні коливання (4500–5000), що свідчить про те, що після десятиліття в ІТ зарплата перестає залежати лише від кількості років. На перший план виходять унікальні навички, ніша та відповідальність.

Далі тут можна зробити багато аналогічних агрегацій, проаналізувати, в який час люди заповнювали форму, хто більше отримує, Business Analyst, чи Data Analyst. Я вас заохочувати не стримувати свій інтерес і покапати ці дані далі :)